pylint/checkers/unicode.py

```diff
@@ -0,0 +1,670 @@
+
+        # If there is no invalid codec already, lets try to determine a BOM,
+        # even so Python currently doesn't like non UTF-8 BOMs ....
+        if invalid_codec_line is None:
+            with node.stream() as stream:
+                codec = extract_codec_from_bom(stream.readline())
+            if self.is_invalid_codec(codec):
+                invalid_codec_line = 1
+
+        if invalid_codec_line is not None:
+            self.add_message(
+                "invalid-unicode-codec",
+                line=invalid_codec_line,
+                confidence=interfaces.HIGH,
+                col_offset=0,
+            )
+
+
+class BidirectionalUnicodeChecker(BaseChecker):
+    """
+    Check Unicode that could be used to hide bad code to humans
+    """
+
+    # Based on:
+    # https://golangexample.com/go-linter-which-checks-for-dangerous-unicode-character-sequences/
+    BAD_UNICODE = [
+        "\u202A",
+        "\u202B",
+        "\u202C",
+        "\u202D",
+        "\u202E",
+        "\u2066",
+        "\u2067",
+        "\u2068",
+        "\u2069",
+        # The following was part of the PEP 672 project:
+        # https://www.python.org/dev/peps/pep-0672/
+        # so the list above might not be complete
+        "\u200f",
+        # We don't use "\u200e" (LEFT to RIGHT MARK)
+        # as this is the default for latin files and can't be used
+        # to hide code
+    ]
+
+    __implements__ = interfaces.IRawChecker
+    priority = -1
+
+    name = "bidirectional_unicode_checker"
+    msgs = {
+        "E1801": (
+            "bidirectional unicode can be used to hide real code execution",
+            "bidirectional-unicode",
+            "bidirectional unicode can be used to hide real code execution",
+        )
+    }
+    options = ()
+
+    def process_module(self, node: nodes.Module) -> None:
+        """
+        Perform the the actual checking by checking module stream
+
+        the module's content is accessible via node.stream() function
+        """
+        with node.stream() as stream:
+            for (lineno, line) in enumerate(stream):
+                line_contains_error: bool = False
+                for dangerous in self.BAD_UNICODE:
+                    # We assume that the file is encoded in UTF-8!
+                    # Otherwise the UTF-16/UTF-32 check would warn!
+                    if (
+                        cached_encode(dangerous, "UTF-8") in line
+                        and not line_contains_error
+                    ):
+                        # Note that we don't add a col_offset on purpose:
+                        #   By definition these unicode characters make it hard
+                        #   and editor depended to see where they are located.
+                        #   A col offset would not help much in this case.
+                        self.add_message(
+                            "bidirectional-unicode",
+                            line=lineno + 1,
```

Comment:
```suggestion
                            node=node,
```

Reply:
Sorry this doesn't work.
We check are in the module there. Pylint will crash as it won't be able to determine the correct line number from the module.

Reply:
Why will it crash? `nodes.Module` should have a `lineno`. In fact with https://github.com/PyCQA/astroid/pull/1262 and `astroid` `2.9.1` a `nodes.Module` will even have ``col_offset``, ``end_lineno``, and ``end_col_offset`` attributes. They don't provide much additional details, but it is better not to lose the context of the `node` as soon as we enter `add_message`.

Reply:
>  E           AssertionError: Pylint analysis failed because of 'Fatal error while checking '/home/daten/Eigene/Programmierung/python/pylint/tests/functional/u/unicode/unicode_bidi_early_return.py'. Please open an issue in our bug tracker so we address this. There is a pre-filled template that you can use in '/home/kound/.cache/pylint/pylint-crash-2021-12-31-15.txt'.'

Content of the file


Issue title:
Crash ``Message E2402 must only provide line, got line=11, node=Module.unicode_bidi_commenting_out(name='unicode_bidi_commenting_out',
                                   doc=('\n'
                                    'Example #1 of trojan unicode, see '
                                    'https://trojansource.codes/\n'
                                    'Taken from '
                                    'https://github.com/nickboucher/trojan-source/tree/main/Python\n'),
                                   file='/home/daten/Eigene/Programmierung/python/pylint/tests/functional/u/unicode/unicode_bidi_commenting_out.py',
                                   path=[ '/home/daten/Eigene/Programmierung/python/pylint/tests/functional/u/unicode/unicode_bidi_commenting_out.py'],
                                   package=False,
                                   pure_python=True,
                                   future_imports=set(),
                                   body=[ <FunctionDef.a_function l.7 at 0x7fbdf9b6bb80>])`` (if possible, be more specific about what made pylint crash)
Content:
When parsing the following file:

<!--
 If sharing the code is not an option, please state so,
 but providing only the stacktrace would still be helpful.
 -->

```python
"""
Example #1 of trojan unicode, see https://trojansource.codes/
Taken from https://github.com/nickboucher/trojan-source/tree/main/Python
"""


def a_function():
    """A simple function"""
    access_level = "user"
    # +1: [bidirectional-unicode]
    if access_level != "none‮⁦":  # Check if admin ⁩⁦' and access_level != 'user
        print("You are an admin.")

```

pylint crashed with a ``InvalidMessageError`` and with the following stacktrace:
```
Traceback (most recent call last):
  File "/home/daten/Eigene/Programmierung/python/pylint/pylint/lint/pylinter.py", line 1072, in _check_files
    self._check_file(get_ast, check_astroid_module, file)
  File "/home/daten/Eigene/Programmierung/python/pylint/pylint/lint/pylinter.py", line 1107, in _check_file
    check_astroid_module(ast_node)
  File "/home/daten/Eigene/Programmierung/python/pylint/pylint/lint/pylinter.py", line 1241, in check_astroid_module
    retval = self._check_astroid_module(
  File "/home/daten/Eigene/Programmierung/python/pylint/pylint/lint/pylinter.py", line 1284, in _check_astroid_module
    checker.process_module(node)
  File "/home/daten/Eigene/Programmierung/python/pylint/pylint/checkers/unicode.py", line 298, in process_module
    self.add_message(
  File "/home/daten/Eigene/Programmierung/python/pylint/pylint/checkers/base_checker.py", line 123, in add_message
    self.linter.add_message(
  File "/home/daten/Eigene/Programmierung/python/pylint/pylint/lint/pylinter.py", line 1566, in add_message
    self._add_one_message(
  File "/home/daten/Eigene/Programmierung/python/pylint/pylint/lint/pylinter.py", line 1468, in _add_one_message
    message_definition.check_message_definition(line, node)
  File "/home/daten/Eigene/Programmierung/python/pylint/pylint/message/message_definition.py", line 107, in check_message_definition
    raise InvalidMessageError(
pylint.exceptions.InvalidMessageError: Message E2402 must only provide line, got line=11, node=Module.unicode_bidi_commenting_out(name='unicode_bidi_commenting_out',
                                   doc=('\n'
                                    'Example #1 of trojan unicode, see '
                                    'https://trojansource.codes/\n'
                                    'Taken from '
                                    'https://github.com/nickboucher/trojan-source/tree/main/Python\n'),
                                   file='/home/daten/Eigene/Programmierung/python/pylint/tests/functional/u/unicode/unicode_bidi_commenting_out.py',
                                   path=[ '/home/daten/Eigene/Programmierung/python/pylint/tests/functional/u/unicode/unicode_bidi_commenting_out.py'],
                                   package=False,
                                   pure_python=True,
                                   future_imports=set(),
                                   body=[ <FunctionDef.a_function l.7 at 0x7fbdf9b6bb80>])

Reply:
Ah, that's what was preventing me from updating all messages to have a node.

We should work on deprecating this `__implements__` stuff and use actual subclassing. Never mind my comment then, this is something to be fixed in another PR.